internal/app/upstream/client.go

```diff
@@ -55,11 +74,177 @@ type Response struct {
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+// NewMockClient creates a mock implementation for testing
+func NewMockClient(
+	ctx context.Context,
+	ldsClient v2.ListenerDiscoveryServiceClient,
+	rdsClient v2.RouteDiscoveryServiceClient,
+	edsClient v2.EndpointDiscoveryServiceClient,
+	cdsClient v2.ClusterDiscoveryServiceClient,
+	callOptions CallOptions) Client {
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}
+}
+
+func (m *client) OpenStream(ctx context.Context, request *v2.DiscoveryRequest) (<-chan *Response, error) {
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case listenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case clusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case routeTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case endpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		return nil, fmt.Errorf("Unsupported request type")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	signal := make(chan *version, 1)
+	signal <- &version{}
+	response := make(chan *Response)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go send(ctx, &wg, request, response, stream, signal, m.callOptions)
+	go recv(ctx, &wg, response, stream, signal)
+
+	go func() {
+		wg.Wait()
+		closeChannel(signal, response)
+	}()
+
+	return response, nil
+}
+
+func send(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	request *v2.DiscoveryRequest,
+	response chan *Response,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					wg.Done()
+					return
+				default:
+					response <- &Response{Err: err}
+					wg.Done()
+					return
+				}
+			}
+		case <-ctx.Done():
+			_ = stream.CloseSend()
+			wg.Done()
+			return
+		}
+	}
+}
+
+func recv(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	response chan *Response,
+	stream grpc.ClientStream,
+	signal chan *version) {
+	for {
+		resp := new(v2.DiscoveryResponse)
+		if err := stream.RecvMsg(resp); err != nil {
+			select {
+			case <-ctx.Done():
+				wg.Done()
+			default:
+				response <- &Response{Err: err}
+			}
+			return
```

Comment:
How come we break out in the error scenario rather than trying to receive future messages?

Reply:
Both [SendMsg](https://github.com/grpc/grpc-go/blob/master/stream.go#L715) and [RecvMsg](https://github.com/grpc/grpc-go/blob/master/stream.go#L734) perform the retries based on the grpc retry params. Since they are already performing the retries, i think we wouldn't benefit from retrying on top of them.

As i have put in `TODO` figuring out grpc retry settings will be a separate PR. When an error happens at this point, the best course would be to open a new stream.